File: platforms/ios/devkit/Sources/NativeblocksDevkit/feature/live/data/LiveRepositoryImpl.swift
import Combine
import Foundation

internal class LiveRepositoryImpl: NSObject, LiveRepository {
    private static let TAG = "LiveRepository"
    private let signalingClient: SocketClient
    private let kitEnvironment: DevKitEnvironment
    private let remoteSession: IRemoteSession

    private let connectionStateSubject = CurrentValueSubject<ConnectionState, Never>(.notConnected)
    private let hotReloadSubject = PassthroughSubject<String, Never>()

    private var streamingToken: String = ""
    private var streamingChannel: String = ""
    private var streamingId: String = ""

    private var cancellables = Set<AnyCancellable>()
    private var orientationChangeListener: OrientationUtils? = nil

    init(
        signalingClient: SocketClient,
        kitEnvironment: DevKitEnvironment,
        remoteSession: IRemoteSession
    ) {
        DevKitLogger.debug(LiveRepositoryImpl.TAG, "init")
        self.signalingClient = signalingClient
        self.kitEnvironment = kitEnvironment
        self.remoteSession = remoteSession
        self.orientationChangeListener = OrientationUtils()
    }

    private var isScreenSharing: Bool { kitEnvironment.screenSharing }

    private var username: String { remoteSession.getUsername() }

    private var uuid: String { remoteSession.getUUID() }

    private var projectId: String {
        JWTUtils.decodedBody(kitEnvironment.apiKey)?["id"] as? String ?? ""
    }

    func start() {
        DevKitLogger.debug(LiveRepositoryImpl.TAG, "start")
        signalingClient.signals().sink(
            receiveCompletion: { completion in
                if case .failure(let error) = completion {
                    DevKitLogger.debug(LiveRepositoryImpl.TAG, "Signals error: \(error.localizedDescription)")
                }
            },
            receiveValue: { [weak self] state in
                self?.handleWebSocketState(state)
            }
        )
        .store(in: &cancellables)

        signalingClient.connect(urlString: remoteSession.getRealtimeEndpoint())
    }

    func end() {
        DevKitLogger.debug(LiveRepositoryImpl.TAG, "end")
        endSession()
    }

    func orientedChange() {
        DevKitLogger.debug(LiveRepositoryImpl.TAG, "orientedChange")
        orientationChangeListener?.register { _ in
            if self.isScreenSharing {
                // NO-OP -> adjustVideoSize()
            }
        }
    }

    func connectionState() -> AnyPublisher<ConnectionState, Never> {
        return connectionStateSubject.eraseToAnyPublisher()
    }

    func hotReload() -> AnyPublisher<String, Never> {
        return hotReloadSubject.eraseToAnyPublisher()
    }

    private func handleWebSocketState(_ state: WebSocketState) {
        switch state {
        case .connecting:
            DevKitLogger.debug(LiveRepositoryImpl.TAG, "WebSocket Connecting")
            connectionStateSubject.send(.connecting)
        case .connected:
            DevKitLogger.debug(LiveRepositoryImpl.TAG, "WebSocket Connected")
            connectionStateSubject.send(.connected)
            startSession()
        case .disconnected:
            DevKitLogger.debug(LiveRepositoryImpl.TAG, "WebSocket Disconnected")
            connectionStateSubject.send(.notConnected)
        case .receiveSignal(let signal):
            handleReceivedSignal(signal)
        case .error(let message, let error):
            DevKitLogger.debug(LiveRepositoryImpl.TAG, "WebSocket Error: \(message), \(error.localizedDescription)")
        //            connectionStateSubject.send(.notConnected)
        }
    }

    private func handleReceivedSignal(_ signal: SignalingModel) {
        DevKitLogger.debug(LiveRepositoryImpl.TAG, "Received signal type: \(signal.type)")

        switch signal {
        case let msg as StartSessionMessage:
            DevKitLogger.debug(LiveRepositoryImpl.TAG, "StartSession acknowledged for client: \(msg.clientId)")
        // SDK doesn't need special handling for StartSession acknowledgment

        case _ as EndSessionMessage:
            DevKitLogger.debug(LiveRepositoryImpl.TAG, "EndSession received")
            streamingToken = ""
            streamingChannel = ""
            streamingId = ""

        case let msg as HotReloadMessage:
            DevKitLogger.debug(LiveRepositoryImpl.TAG, "HotReload received")
            let frameRoute = msg.frameRoute ?? ""
            hotReloadSubject.send(frameRoute)

        case _ as AvailableTargetsMessage:
            DevKitLogger.debug(LiveRepositoryImpl.TAG, "AvailableTargets received (Studio-only message)")
        // SDK doesn't need to handle AvailableTargets (Studio-only message)

        case let msg as LogEventMessage:
            DevKitLogger.debug(LiveRepositoryImpl.TAG, "Log event received from Studio: \(msg.message)")
        // SDK doesn't need to handle log events from Studio (optional for bidirectional logging)

        default:
            DevKitLogger.debug(LiveRepositoryImpl.TAG, "Unknown signal type")
        }
    }

    private func startSession() {
        DevKitLogger.debug(LiveRepositoryImpl.TAG, "startSession")
        if isScreenSharing {
            // NO-OP -> adjustVideoSize()
        }

        let message = StartSessionMessage(
            clientType: .SDK,
            clientId: uuid,
            roomId: projectId,
            username: username,
            deviceName: DeviceInfo.getDeviceName(),
            hasSharingScreen: isScreenSharing
        )
        signalingClient.send(message: message)
    }

    private func endSession() {
        DevKitLogger.debug(LiveRepositoryImpl.TAG, "endSession")
        let message = EndSessionMessage(
            clientId: uuid,
            roomId: projectId
        )
        signalingClient.send(message: message)

        streamingToken = ""
        streamingChannel = ""
        streamingId = ""

        connectionStateSubject.send(.notConnected)
        if isScreenSharing {
            // NO-OP -> endStreaming()
        }
        signalingClient.disconnect()
    }

    func cleanUp() {
        DevKitLogger.debug(LiveRepositoryImpl.TAG, "cleanUp")
        orientationChangeListener?.unregister()
        orientationChangeListener = nil
        cancellables.forEach { $0.cancel() }
        cancellables.removeAll()
        signalingClient.disconnect()
    }
}

File: platforms/android/devkit/devkit/src/main/java/io/nativeblocks/devkit/feature/live/data/LiveRepositoryImpl.kt
package io.nativeblocks.devkit.feature.live.data

import android.content.Context
import android.os.Build
import io.nativeblocks.devkit.feature.live.domain.model.ConnectionState
import io.nativeblocks.devkit.feature.live.domain.repository.LiveRepository
import io.nativeblocks.devkit.session.DevKitEnvironment
import io.nativeblocks.devkit.session.IRemoteSession
import io.nativeblocks.devkit.util.DevKitLogger
import io.nativeblocks.devkit.util.JWT
import io.nativeblocks.devkit.util.registerOrientationChange
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.catch

internal class LiveRepositoryImpl(
    private val context: Context,
    private val signalingClient: SocketClient,
    private val kitEnvironment: DevKitEnvironment,
    private val remoteSession: IRemoteSession,
) : LiveRepository {

    companion object {
        private val TAG = LiveRepository::class.java.simpleName
    }

    private val _connectionState = MutableStateFlow(ConnectionState.NotConnected)
    private val _hotReload = MutableSharedFlow<String>()

    private val isScreenSharing
        get() = kitEnvironment.screenSharing

    private val username: String
        get() = remoteSession.getUsername()
    private val uuid: String
        get() = remoteSession.getUUID()
    private val projectId: String by lazy {
        JWT.decodedBody(kitEnvironment.apiKey)?.getString("id") ?: ""
    }

    private var streamingToken: String = ""
    private var streamingChannel: String = ""
    private var streamingId: String = ""

    private fun getDeviceName(): String {
        val manufacture = Build.MANUFACTURER
        val model = Build.MODEL
        return if (model.startsWith(manufacture)) {
            model
        } else {
            "$manufacture $model"
        }
    }

    override suspend fun start() {
        DevKitLogger.d(TAG, "registerOrientationChange")
        DevKitLogger.d(TAG, "start")
        startSignal()
    }

    override suspend fun end() {
        DevKitLogger.d(TAG, "end")
        endSession()
    }

    override suspend fun orientedChange() {
        DevKitLogger.d(TAG, "orientedChange")
        registerOrientationChange(context).collect {
            if (isScreenSharing) {
                // NO-OP -> adjustVideoSize()
            }
        }
    }

    override fun connectionState() = _connectionState.asStateFlow()

    override fun hotReload() = _hotReload.asSharedFlow()

    private suspend fun startSignal() {
        DevKitLogger.d(TAG, "startSignal")
        signalingClient.signals(remoteSession.getRealtimeEndpoint()).catch {
            DevKitLogger.e(TAG, "Signals error: ${it.message}")
        }.collect { state ->
            when (state) {
                is WebSocketState.ReceiveSignal -> {
                    DevKitLogger.d(TAG, "WebSocket ReceiveSignal ${state.signal}")
                    when (val signal = state.signal) {
                        is StartSessionMessage -> {
                            DevKitLogger.d(TAG, "StartSession acknowledged")
                        }

                        is EndSessionMessage -> {
                            streamingToken = ""
                            streamingChannel = ""
                            streamingId = ""
                        }

                        is HotReloadMessage -> {
                            val frameRoute = signal.frameRoute ?: ""
                            _hotReload.emit(frameRoute)
                        }

                        is AvailableTargetsMessage -> {
                            // SDK doesn't need to handle AvailableTargets (Studio-only message)
                            DevKitLogger.d(TAG, "AvailableTargets received (Studio-only message)")
                        }

                        is LogEventMessage -> {
                            // SDK doesn't need to handle log events from Studio (optional for bidirectional logging)
                            DevKitLogger.d(TAG, "Log event received from Studio: ${signal.message}")
                        }
                    }
                }

                WebSocketState.Connecting -> {
                    DevKitLogger.d(TAG, "WebSocket Connecting")
                    _connectionState.emit(ConnectionState.Connecting)
                }

                WebSocketState.Connected -> {
                    DevKitLogger.d(TAG, "WebSocket Connected")
                    _connectionState.emit(ConnectionState.Connected)
                    startSession()
                }

                WebSocketState.Disconnected -> {
                    _connectionState.emit(ConnectionState.NotConnected)
                    DevKitLogger.d(TAG, "WebSocket Disconnected")
                }

                is WebSocketState.Error -> {
                    DevKitLogger.e(TAG, "WebSocket Error: ${state.message}", state.error)
                    _connectionState.emit(ConnectionState.NotConnected)
                }
            }
        }
    }

    private suspend fun startSession() {
        DevKitLogger.d(TAG, "startSession")
        if (isScreenSharing) {
            // NO-OP -> adjustVideoSize()
        }
        signalingClient.send(
            StartSessionMessage(
                clientType = SignalingClientType.SDK,
                clientId = uuid,
                roomId = projectId,
                username = username,
                deviceName = getDeviceName(),
                hasSharingScreen = isScreenSharing
            )
        )
    }

    private suspend fun endSession() {
        DevKitLogger.d(TAG, "endSession")
        signalingClient.send(
            EndSessionMessage(
                clientId = uuid,
                roomId = projectId
            )
        )
        streamingToken = ""
        streamingChannel = ""
        streamingId = ""
        _connectionState.emit(ConnectionState.NotConnected)
        if (isScreenSharing) {
            // NO-OP -> endStreaming()
        }
    }
}